<?xml version="1.0"?>
<launch>
    <!-- Set global use_sim_time parameter -->
    <arg name="use_sim_time" default="true"/>
    
    <!-- Launch group for carter1 -->
    <group>
        <!-- SLAM Toolbox -->
        <node pkg="slam_toolbox" exec="sync_slam_toolbox_node" name="slam_toolbox" output="screen">
            <param name="use_sim_time" value="$(var use_sim_time)"/>
            <param name="odom_frame" value="odom"/>
            <param name="base_frame" value="base_link"/>
            <param name="map_frame" value="carter1/map"/>
            <param name="message_queue_size" value="2000"/>
            <param name="minimum_time_interval" value="0.5"/>
            <param name="transform_timeout" value="0.2"/>
            <param name="tf_buffer_duration" value="30.0"/>

            <!-- Remappings -->
            <remap from="scan" to="/carter1/front_3d_lidar/lidar_points"/>
            <remap from="odom" to="/carter1/chassis/odom"/>
            <remap from="tf" to="/carter1/tf"/>
            <remap from="tf_static" to="/carter1/tf_static"/>
            <remap from="map" to="/carter1/map"/>
        </node>

        <!-- Frontier Explorer for carter1 -->
        <node pkg="frontier_explorer" exec="frontier_explorer" 
              name="frontier_explorer_carter1" 
              output="screen">
            <param name="use_sim_time" value="$(var use_sim_time)"/>
            <param name="robot_namespace" value="carter1"/>
            <param name="map_frame" value="carter1/map"/>
            <remap from="tf" to="/carter1/tf"/>
            <remap from="tf_static" to="/carter1/tf_static"/>
        </node>
    </group>

    <!-- Launch group for carter2 -->
    <group>
        <!-- SLAM Toolbox -->
        <node pkg="slam_toolbox" exec="sync_slam_toolbox_node" name="slam_toolbox_carter2" output="screen">
            <param name="use_sim_time" value="$(var use_sim_time)"/>
            <param name="odom_frame" value="odom"/>
            <param name="base_frame" value="base_link"/>
            <param name="map_frame" value="carter2/map"/> 
            <param name="message_queue_size" value="2000"/>
            <param name="minimum_time_interval" value="0.5"/>
            <param name="transform_timeout" value="0.2"/>
            <param name="tf_buffer_duration" value="30.0"/>

            <!-- Remappings -->
            <remap from="scan" to="/carter2/front_3d_lidar/lidar_points"/>
            <remap from="odom" to="/carter2/chassis/odom"/>
            <remap from="tf" to="/carter2/tf"/>
            <remap from="tf_static" to="/carter2/tf_static"/>
            <remap from="map" to="/carter2/map"/>
        </node>

        <!-- Frontier Explorer for carter2 -->
        <node pkg="frontier_explorer" exec="frontier_explorer" 
              name="frontier_explorer_carter2" 
              output="screen">
            <param name="use_sim_time" value="$(var use_sim_time)"/>
            <param name="robot_namespace" value="carter2"/>
            <param name="map_frame" value="carter2/map"/>
            <remap from="tf" to="/carter2/tf"/>
            <remap from="tf_static" to="/carter2/tf_static"/>
        </node>
    </group>
</launch>